File: src/primitives.ts
import { Effect } from "effect";
import { Either } from "effect";
import type { Parser } from "./types";
import { success, error, updatePosition } from "./utils";

export const char = (expected: string): Parser<string> => ({
  parse: (input) =>
    Effect.succeed(() => {
      if (input.text.charAt(0) === expected) {
        const remaining = {
          text: input.text.slice(1),
          position: updatePosition(input, expected),
        };
        return Either.right(success(expected, remaining));
      }
      return Either.left(
        error(
          `Expected '${expected}', got '${input.text.charAt(0)}'`,
          input.position
        )
      );
    }),
});

export const string = (expected: string): Parser<string> => ({
  parse: (input) =>
    Effect.succeed(() => {
      if (input.text.startsWith(expected)) {
        const remaining = {
          text: input.text.slice(expected.length),
          position: updatePosition(input, expected),
        };
        return Either.right(success(expected, remaining));
      }
      return Either.left(
        error(
          `Expected "${expected}", got "${input.text.slice(
            0,
            expected.length
          )}"`,
          input.position
        )
      );
    }),
});

export const regex = (
  pattern: RegExp,
  description: string
): Parser<string> => ({
  parse: (input) =>
    Effect.succeed(() => {
      const match = pattern.exec(input.text);
      if (match && match.index === 0) {
        const matched = match[0];
        const remaining = {
          text: input.text.slice(matched.length),
          position: updatePosition(input, matched),
        };
        return Either.right(success(matched, remaining));
      }
      return Either.left(error(`Expected ${description}`, input.position));
    }),
});

File: parso/primitives.py
import re
from toolz import curry
from .types import ParserInput, Result, Parser, Success, Ok, Err
from .utils import update_position, success, error

@curry
def char(expected: str) -> Parser[str]:
    def parse_fn(input: ParserInput) -> Result[Success[str]]:
        if input.text and input.text[0] == expected:
            remaining = ParserInput(
                text=input.text[1:],
                position=update_position(input, expected)
            )
            return Ok(success(expected, remaining))
        return Err(error(f"Expected '{expected}', got '{input.text[0] if input.text else 'EOF'}'", input.position))
    return Parser(parse_fn)

@curry
def string(expected: str) -> Parser[str]:
    def parse_fn(input: ParserInput) -> Result[Success[str]]:
        if input.text.startswith(expected):
            remaining = ParserInput(
                text=input.text[len(expected):],
                position=update_position(input, expected)
            )
            return Ok(success(expected, remaining))
        return Err(error(f'Expected "{expected}", got "{input.text[:len(expected)]}"', input.position))
    return Parser(parse_fn)

@curry
def regex(pattern: re.Pattern, description: str) -> Parser[str]:
    def parse_fn(input: ParserInput) -> Result[Success[str]]:
        match = pattern.match(input.text)
        if match:
            matched = match.group(0)
            remaining = ParserInput(
                text=input.text[len(matched):],
                position=update_position(input, matched)
            )
            return Ok(success(matched, remaining))
        return Err(error(f"Expected {description}", input.position))
    return Parser(parse_fn)
